pkg/il/evaluator/evaluator.go

```diff
@@ -32,9 +34,17 @@ import (
 // IL is an implementation of expr.Evaluator that also exposes specific methods.
 // Specifically, it can listen to config change events.
 type IL struct {
+	cacheSize       int
+	attrContext     *attrContext
+	attrContextLock sync.RWMutex
```

Comment:
not a blocker, but something to consider is just embedding `sync.RWMutex`  anonymously on `IL`:

```
type IL struct {
        sync.RWMutex
...
```

That would allow you to have:

```
e.RLock()
...
e.RUnlock()
```

instead of: `e.attrContextLock.RUnlock()`, etc.

Of course, that implies a slightly different semantic (is the fMap protected too?)

Reply:
The lock specifically protects the attrContext. I'd prefer to keep it that way to reduce to scope of locking (and contention). We can refactor as needed when/if we make a major change to the IL struct.